velox/functions/sparksql/tests/StringTest.cpp

```diff
@@ -497,5 +522,105 @@ TEST_F(StringTest, left) {
   EXPECT_EQ(left("da\u6570\u636Eta", 30), "da\u6570\u636Eta");
 }
 
+void StringTest::testTranslateFlatVector(
+    exec::ExprSet& exprSet,
+    const translate_input_test_t& inputData,
+    const std::optional<bool>& asciiEncoding) {
+  auto stringVector = makeFlatVector<StringView>(inputData.size());
+  for (int i = 0; i < inputData.size(); i++) {
+    stringVector->set(i, StringView(inputData[i].first));
+  }
+  if (asciiEncoding.has_value()) {
+    stringVector->setAllIsAscii(asciiEncoding.value());
+  }
+  auto result = evaluate(exprSet, makeRowVector({stringVector}));
+  FlatVector<StringView>* flatResult = result->asFlatVector<StringView>();
+  for (int32_t i = 0; i < inputData.size(); ++i) {
+    ASSERT_EQ(flatResult->valueAt(i), inputData[i].second);
+  }
+}
+
+void StringTest::testTranslateAllFlatVector(
+    exec::ExprSet& exprSet,
+    const translate_all_input_test_t& inputData,
+    const std::vector<std::optional<bool>>& asciiEncodings) {
+  auto stringVector = makeFlatVector<StringView>(inputData.size());
+  auto matchVector = makeFlatVector<StringView>(inputData.size());
+  auto replaceVector = makeFlatVector<StringView>(inputData.size());
+  for (int i = 0; i < inputData.size(); i++) {
+    stringVector->set(i, StringView(std::get<0>(inputData[i].first)));
+    matchVector->set(i, StringView(std::get<1>(inputData[i].first)));
+    replaceVector->set(i, StringView(std::get<2>(inputData[i].first)));
+  }
+  if (asciiEncodings[0].has_value()) {
+    stringVector->setAllIsAscii(asciiEncodings[0].value());
+  }
+  if (asciiEncodings[1].has_value()) {
+    matchVector->setAllIsAscii(asciiEncodings[1].value());
+  }
+  if (asciiEncodings[2].has_value()) {
+    replaceVector->setAllIsAscii(asciiEncodings[2].value());
+  }
+  auto result = evaluate(
+      exprSet, makeRowVector({stringVector, matchVector, replaceVector}));
+  FlatVector<StringView>* flatResult = result->asFlatVector<StringView>();
+  for (int32_t i = 0; i < inputData.size(); ++i) {
+    ASSERT_EQ(flatResult->valueAt(i), inputData[i].second);
+  }
+}
+
+TEST_F(StringTest, translate) {
+  EXPECT_EQ(translate("ab[cd]", "[]", "##"), "ab#cd#");
+  EXPECT_EQ(translate("ab[cd]", "[]", "#"), "ab#cd");
+  EXPECT_EQ(translate("ab[cd]", "[]", "#@$"), "ab#cd@");
+  EXPECT_EQ(translate("ab[cd]", "[]", "  "), "ab cd ");
+  EXPECT_EQ(translate("ab\u2028", "\u2028", "\u2029"), "ab\u2029");
+  EXPECT_EQ(translate("abcabc", "a", "\u2029"), "\u2029bc\u2029bc");
+  EXPECT_EQ(translate("abc", "", ""), "abc");
+  EXPECT_EQ(translate("translate", "rnlt", "123"), "1a2s3ae");
+  EXPECT_EQ(translate("translate", "rnlt", ""), "asae");
+  EXPECT_EQ(translate("abcd", "aba", "123"), "12cd");
+  // Test null input.
+  EXPECT_EQ(translate("abc", std::nullopt, "\u2029"), std::nullopt);
+  EXPECT_EQ(translate("abc", "\u2028", std::nullopt), std::nullopt);
+  EXPECT_EQ(translate(std::nullopt, "\u2028", "\u2029"), std::nullopt);
+
+  // Test with flat vector.
```

Comment:
This comment as well as method name testTranslateFlatVector suggests that there is something special about flat vectors that requires specific testing. I believe this is not the case. In fact, the function implementation is agnostic to vector encoding as it receives values one row at at time and doesn't have any handing of encodings. The fragment handles the encoding.

We do need to test fast ASCII paths as well as cases where we have a mix of all-ASCII and non-all-ASCII batches of data.

Would it be possible to refactor this tests to make it clear what is being tested?

Reply:
I just refactored the test code. And the encoding setting for test batch is removed. It may look more clear to reader now. 

BTW, in the latest implementation for translate, I introduced `std::optional` for the two dictionaries and removed the two flags that used for marking whether they are built.